### PR TITLE
Backport recipe map from dev-3.x (closes #869, #874, #884, #886, #893, #894, #898, #901, #903, #908, #918)

### LINKED ISSUE #869: recipe-map calculation breaks autopkg if a recipe exists with a syntax error
If you have a broken recipe in any of the repos in the cache, recipe-map will fail to build, and autopkg cannot be run. It's not very easy to find which recipe/repo is the culprit, so remediation can be painful. I have reproduced by forking a repo I didn't have before and then adding it (twice, due to #868).

**Preferences contents**
Same as in #868

**AutoPkg output**
```
% autopkg repo-add grahampugh/macprince-recipes
Attempting git clone for https://github.com/grahampugh/macprince-recipes...

Adding /Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.grahampugh.macprince-recipes to RECIPE_SEARCH_DIRS...
Updating recipe map...
Updated search paths:
  '.'
  '~/Library/AutoPkg/Recipes'
  '/Library/AutoPkg/Recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.rtrouton-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.ahousseini-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.andredb90-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.apizz-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.blackthroat-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.bochoven-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.cgerke-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.crystalllized-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.dataJAR-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.drewdiver-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.eth-its.autopkg-mac-recipes-yaml'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.faumac-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.gerardkok-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.hansen-m-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.hjuutilainen-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.homebysix-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.its-unibas-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.jbaker10-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.jlehikoinen-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.joshua-d-miller-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.justinrummel-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.keeleysam-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.killahquam-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.MLBZ521-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.moofit-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.n8felton-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.neilmartin83-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.nmcspadden-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.novaksam-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.nstrauss-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.nzmacgeek-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.paul-cossey-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.peshay-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.peterkelm-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.precursorca-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.scriptingosx-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.sebtomasi-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.smithjw-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.tbridge-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.wardsparadox-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.jss-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.davidbpirie-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.jaharmi-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.kitzy-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.erikng-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.kevinmcox-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.swy-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.jleggat-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.zentral-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.grahampugh.jamf-upload'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.grahampugh-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.autopkg.tallfunnyjew-recipes'
  '/Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.grahampugh.macprince-recipes'

% autopkg repo-update grahampugh/macprince-recipes
Attempting git pull for /Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.grahampugh.macprince-recipes...
Already up to date.

Recalculating the recipe map...
WARNING: plist error for /Users/gpugh/Library/AutoPkg/RecipeRepos/com.github.grahampugh.macprince-recipes/DeepL/DeepL.download.recipe: Invalid file
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 2715, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 2711, in main
    return subcommands[verb]["function"](argv)
  File "/usr/local/bin/autopkg", line 818, in repo_update
    calculate_recipe_map()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 523, in calculate_recipe_map
    write_recipe_map_to_disk()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 557, in write_recipe_map_to_disk
    json.dump(
  File "/Library/AutoPkg/Python3/Python.framework/Versions/3.10/lib/python3.10/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/Library/AutoPkg/Python3/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/Library/AutoPkg/Python3/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Library/AutoPkg/Python3/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py", line 353, in _iterencode_dict
    items = sorted(dct.items())
TypeError: '<' not supported between instances of 'NoneType' and 'str'
```

**Expected behavior**
Either an invalid recipe should be ignored (not added to the map) or a valid warning should be outputted that says which recipe is invalid, so that it can be easily remediated.

**Version (please complete the following information):**
 - OS version: 14.0b3
- AutoPkg Version: 2.8 RC

### PR DESCRIPTION
**Pull request by smithjw**
## What this is

This PR backports the recipe map functionality from `dev-3.x` to `dev-2.x`, picking up the maintainer suggestion in #1026 that the existing PR be re-implemented from the dev-3.x state rather than rebased.

The recipe map is an on-disk JSON cache of every recipe and override on the system, indexed by identifier and shortname. Recipe resolution becomes O(1) instead of walking every configured `RECIPE_SEARCH_DIRS` entry on each invocation — the dominant cost for users with many recipe repos. The cache lives at `~/Library/AutoPkg/recipe_map.json` by default.

## Why this approach

I worked through the recipe-map commits on `dev-3.x`, but there are enough structural differences (the `common.py`/`prefs.py` split, the `Recipe` class refactor in `autopkglib/recipes/`) that cherry-picking would have produced a tangled history with conflicts in nearly every commit. Instead I synthesised a clean port targeted at the dev-2.x layout, with one logical commit per layer. Each commit builds and the full test suite passes on its own — easier to review, easier to revert if needed.

While I was in the area, I picked up every open recipe-map issue I could find in the tracker so this PR closes a meaningful set of bugs rather than introducing a fresh feature on top of known issues.

## What changed for users

- New `autopkg generate-recipe-map` subcommand to explicitly build and persist the map. Useful for fresh installs and CI/CD pipelines that want to pay the scan cost up front.
- The map auto-rebuilds on first use when it's missing or invalid — fresh installs and ephemeral runners "just work" without a manual bootstrap step. (This is a UX divergence from dev-3.x, which printed an error and exited.)
- `repo-add`, `repo-delete`, `repo-update`, `make-override` and `new-recipe` keep the map in sync automatically. `repo-update` only rebuilds when `git pull` actually changed `HEAD`, so no-op updates don't incur the scan cost.
- CLI `--search-dir` / `--override-dir` flags still scope resolution to exactly the supplied dirs when they differ from prefs. Preserves the dev-2.x "recipes in `~/Library/AutoPkg/Recipes` override installed repos" contract that #886 / #908 / #918 reported as broken in earlier recipe-map iterations.
- `find_processor_path` consults the map first, with a one-shot rebuild that pulls in the current working directory when a shared processor is missing from the persisted map (#894).
- Trust info verification (`verify-trust-info`, `update-trust-info`) correctly classifies override files passed by path (#903).
- Map location is configurable via the `RECIPE_MAP_PATH` preference or `AUTOPKG_RECIPE_MAP_PATH` env var, with env taking precedence (#901).
- Escape hatch: set `DISABLE_RECIPE_MAP=1` (or the env var) to bypass the map and fall back to the legacy on-disk scanners. No code change required if anyone hits a bug in production.

## Issues closed

`#869`, `#874`, `#884`, `#886`, `#893`, `#894`, `#898`, `#901`, `#903`, `#908`, `#918`. Each has at least one regression test pinned to it in `Code/tests/test_autopkg_recipe_map_regressions.py` so future changes that break the behaviour fail in CI with a direct pointer back to the original report.

## How it's structured

8 progressive commits on top of `dev-2.x`. Each commit passes the full test suite on its own:

1. **Pre-port refactor** — consolidate recipe helpers into `autopkglib`. No functional change.
2. **Recipe map core** — `globalRecipeMap`, `calculate_recipe_map`, `read_recipe_map`, lookup helpers, persistence.
3. **Wire into verbs** — integrate the map into `Code/autopkg`'s verbs and `find_recipe`.
4. **Initial test coverage** — 40 unit tests for the core.
5. **Review blockers and known bugs** — addresses every blocker from a code review pass plus the bug fixes for the issues above.
6. **Regression tests** — 38 tests pinned to specific upstream issues / review findings.
7. **Review findings + CHANGELOG** — addresses a follow-up review pass focused on hardening, plus the user-facing CHANGELOG entry.
8. **Post-review cleanup** — simplifications, hot-path perf fix, tightened docs and lint compliance.

## Implementation notes worth flagging

- **Atomic writes.** The persisted map goes through `tempfile.mkstemp` + `os.replace` so concurrent autopkg invocations can't observe a half-written file.
- **Schema version.** Persisted files carry a `schema_version` field; future format changes can trigger a clean rebuild rather than reading stale data.
- **YAML loader hardened.** `recipe_from_file` switched from `yaml.FullLoader` to `yaml.safe_load`. Recipe documents are always plain mappings of primitives so no legitimate recipe is affected, but the recipe map causes every `.recipe.yaml` in every search dir to be parsed on every map build/lookup — much wider attack surface than before this PR. `safe_load` closes that.
- **Root-redirect warning.** Redirecting the map path via env var or pref while running as root now emits a prominent log warning (`SECURITY WARNING: ...`). Ops teams running `sudo autopkg` should strip `AUTOPKG_*` from `env_keep` if they don't want unprivileged callers to influence where autopkg writes.
- **Tests are hermetic.** A shared test fixture redirects `DEFAULT_RECIPE_MAP` to a per-test tempdir and resets `globalRecipeMap`, so the suite never touches the developer's real `~/Library/AutoPkg`. Suite runtime stays at ~28 seconds.

### Automated checks (5 minutes)

```sh
/usr/local/autopkg/python Scripts/run_tests.py
# Expected: 817 tests, 12 skipped, 0 failures.
```

The 78 new tests live in `Code/tests/test_autopkg_recipe_map.py` (40 unit tests) and `Code/tests/test_autopkg_recipe_map_regressions.py` (38 regression tests). Each regression test class cites the issue number it covers in its docstring.

### Hands-on scenarios (~20 minutes)

The most important scenarios to walk through manually, since they cover the user-visible behaviour changes:

| # | Scenario | Validates |
|---|----------|-----------|
| 1 | `rm -f ~/Library/AutoPkg/recipe_map.json && autopkg repo-add recipes` — map auto-creates, no error. | #884, #893, #898 |
| 2 | `autopkg generate-recipe-map` from a fresh install. Persists to disk; reports counts. | New verb |
| 3 | A recipe in `~/Library/AutoPkg/Recipes` with the same identifier as a stock recipe. `autopkg info <id>` resolves to the local copy. | #886, #908, #918 |
| 4 | `autopkg info -d /tmp/scoped-dir <id>` — CLI dir wins over the map. | #886, #908, #918 |
| 5 | A shared-processor recipe in cwd. First reference triggers a rebuild and resolves; subsequent misses don't trigger another rebuild. | #894 |
| 6 | `autopkg update-trust-info ./Overrides/<file>.yaml` — file passed by path, classified as override via the map. | #903 |
| 7 | `autopkg repo-update all` on a fully-up-to-date tree — map mtime unchanged. | Perf |
| 8 | A syntactically-invalid recipe in a search dir. `generate-recipe-map` emits a warning, doesn't crash. | #869 |
| 9 | A pre-v1 map file (no `overrides-identifiers` key). `autopkg info` triggers an auto-rebuild rather than `KeyError`. | #874 |
| 10 | `RECIPE_MAP_PATH` pref or `AUTOPKG_RECIPE_MAP_PATH` env var redirects the persisted location; env wins. | #901 |
| 11 | `AUTOPKG_DISABLE_RECIPE_MAP=1 autopkg info <id>` — falls back to on-disk scan, no map references in log. | Escape hatch |
| 12 | Read-only target directory — write fails with actionable warning, no `.tmp` turd left behind. | Atomic write |

Happy to provide more detailed reproduction scripts for any of these on request.

### If something goes wrong

The CLI surface is additive. If a user hits a recipe-map bug in production:

1. **Bypass**: set `AUTOPKG_DISABLE_RECIPE_MAP=1` or the `DISABLE_RECIPE_MAP` pref to fall back to the dev-2.x on-disk scanners. No code change required.
2. **Rebuild**: delete the map and run `autopkg generate-recipe-map`.
3. **Revert**: the 8 commits are self-contained. Reverting the merge restores pre-backport behaviour exactly.

## Outstanding things

- Marked as draft because I'd appreciate maintainer eyes on the auto-create UX divergence from dev-3.x before this lands. Happy to switch back to the original "exit on missing map" if you'd prefer that to stay consistent with how dev-3.x behaves.
- I'm leaving the `find_recipe_by_identifier` alias in `autopkglib` for backwards compat with anything outside this repo that imports it. Nothing inside the tree uses it any more; it can come out in a follow-up if the maintainer prefers.

Let me know if there's anything else you'd like me to cover, restructure, or back out. Thanks for the suggestion to re-implement from dev-3.x in #1026 — it produced a much cleaner result than rebasing the original would have.

### CODE REVIEW

**Comment by homebysix** [comment]
Given the speed between the suggestion and PR submission I have to assume AI assistance was used. Can you share some details about what your planning approach was, what prompts you used, what AI model and effort, and any adversarial reviews or testing already performed?

File: Code/autopkg [hunk 1]
```diff
@@ -68,6 +82,25 @@ _ = f"""{sys.version_info.major} It looks like you're running the autopkg tool w
...
+# process so pathological multi-miss runs (e.g. `autopkg run -l typos`
+# against a list of non-existent identifiers) don't trigger N full
+# rebuilds for N misses.
+#
+# Test fixtures reset this in setUp via ``_set_locate_recipe_rebuild_attempted``
+# so each test starts from a clean state.
+_locate_recipe_rebuild_attempted = False
+
+
+def _set_locate_recipe_rebuild_attempted(value: bool) -> None:
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] This setter exposes a test-only hook as production API — per the comment at line 91 its sole purpose is letting tests reset the latch. Two production callers (`locate_recipe`, `find_processor_path`) share the flag for unrelated reasons. Consider wrapping both call sites in a `_try_cwd_rebuild_once()` helper that tests can patch directly, removing the need for a module-level mutable + public setter.

File: Code/autopkg [hunk 2]
```diff
@@ -113,79 +146,131 @@ def builds_a_package(recipe) -> bool:
...
+
+    Empty-list and ``None`` semantics: both are treated as "not supplied"
+    — the pref baseline is used. Callers that want to restrict resolution
+    to an empty scope should not call this function; call
+    ``find_recipe_by_identifier_on_disk`` / ``find_recipe_by_name_on_disk``
+    directly instead."""
+    # Caller-supplied dirs that differ from prefs → scope to those dirs
+    # only. The map is pref-scoped and would silently include/exclude the
+    # wrong set of dirs, so we bypass it entirely in that case.
+    effective_search = search_dirs if search_dirs else get_search_dirs()
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] `effective_search` / `effective_override` are computed unconditionally but only used after the `if use_map` early-return. On every map hit the extra `get_search_dirs()` / `get_override_dirs()` calls are wasted. Move the assignments into the miss branch:

```python
scan_dirs: list[str] = []
if not skip_overrides:
    scan_dirs.extend(override_dirs or get_override_dirs())
scan_dirs.extend(search_dirs or get_search_dirs())
```

File: Code/autopkg [hunk 3]
```diff
@@ -745,8 +844,13 @@ Example: '%prog repo-add recipes'
         log_err("Need at least one recipe repo URL!")
         return -1
 
+    # Read the existing map (if any) before modifying prefs so that the
+    # rebuild below sees a consistent baseline.
+    read_recipe_map(allow_continuing=True)
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] This `read_recipe_map(allow_continuing=True)` is wasted work. The subsequent `calculate_recipe_map()` at line 880 clears `globalRecipeMap` and re-reads prefs from disk — it doesn't consult the in-memory state loaded here. The "consistent baseline" comment doesn't match behavior; the baseline is immediately discarded.

File: Code/autopkg [hunk 4]
```diff
@@ -788,8 +900,13 @@ def repo_delete(argv):
         log_err("Need at least one recipe repo path or URL!")
         return -1
 
+    # Read the existing map (if any) before mutating prefs so we can
+    # rebuild it cleanly below.
+    read_recipe_map(allow_continuing=True)
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] Same as `repo-add`: this `read_recipe_map` call is discarded by the following `calculate_recipe_map()` rebuild at line 946.

File: Code/autopkg [hunk 5]
```diff
@@ -1501,24 +1700,81 @@ class TrustVerificationError(AutoPackagerError):
 
 
 def recipe_from_external_repo(recipe_path):
```
Comment (homebysix) [inline Code/autopkg:1689]: [claude-opus-4-7] This reimplements the path-separator-anchored prefix check that `_path_under_dirs` (line 165) already provides. Can collapse to:

```python
recipe_repos = get_pref("RECIPE_REPOS") or {}
return bool(recipe_repos) and _path_under_dirs(recipe_path, list(recipe_repos))
```

File: Code/autopkg [hunk 6]
```diff
@@ -1970,6 +2226,9 @@ def make_override(argv):
         with open(override_file, "wb") as f:
             plistlib.dump(plist_serializer(override_dict), f)
     log(f"Override file saved to {override_file}")
+    # Refresh the recipe map so the newly created override is discoverable
+    # immediately without a separate `generate-recipe-map` step.
+    calculate_recipe_map()
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] `make_override` triggers a full map rebuild for a single-file add. `calculate_recipe_map()` re-walks every repo in `RECIPE_SEARCH_DIRS` and every override dir just to index one new override. Consider reading the persisted map, inserting the new entry into the four sub-dicts, and re-persisting.

File: Code/autopkg [hunk 7]
```diff
@@ -2623,6 +2882,60 @@ def audit(argv):
...
+    options, _arguments = common_parse(parser, argv)
+
+    extra_search_dirs = list(options.search_dirs or [])
+    extra_override_dirs = list(options.override_dirs or [])
+
+    # Single rebuild call covering configured prefs + any CLI-supplied
+    # extras. persist=True ensures the result is written to disk even
+    # when extras were supplied, which is the whole point of this verb
+    # (CI/CD bootstrap).
+    calculate_recipe_map(
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] `persist=True` here overrides the `persist=None` guard in `calculate_recipe_map` that was designed to keep `extra_*` dirs out of `recipe_map.json` (see docstring at `autopkglib/__init__.py:720-732`). When `autopkg generate-recipe-map --search-dir /tmp/foo` is run, those transient paths get baked into the persisted map and pollute every later pref-scoped invocation. Consider rebuilding without extras for the persisted pass, or refusing `--persist` when extras are supplied.

File: Code/autopkg [hunk 8]
```diff
@@ -2684,6 +2997,13 @@ def new_recipe(argv):
         log(f"Saved new recipe to {filename}")
     except Exception as err:
         log_err(f"Failed to write recipe: {err}")
+        return
+
+    # Refresh the recipe map so the newly created recipe is discoverable.
+    # We include the recipe's directory as an extra search dir in case the
+    # caller wrote to somewhere outside the configured RECIPE_SEARCH_DIRS.
+    recipe_dir = os.path.dirname(os.path.abspath(filename)) or "."
+    calculate_recipe_map(extra_search_dirs=[recipe_dir])
```
Comment (homebysix) [inline Code/autopkg]: [claude-opus-4-7] Same full-rebuild pattern as `make_override`. `calculate_recipe_map(extra_search_dirs=[recipe_dir])` rebuilds the whole map to add one recipe; an incremental update would avoid the cost.

File: Code/autopkglib/__init__.py [hunk 9]
```diff
@@ -400,6 +519,529 @@ def find_recipe_by_identifier(identifier, search_dirs) -> str | None:
...
+        "shortnames",
+        "overrides",
+        "overrides-identifiers",
+    ):
+        globalRecipeMap.setdefault(sub, {}).clear()
+
+    extra_search_dirs = list(extra_search_dirs or [])
+    extra_override_dirs = list(extra_override_dirs or [])
+
+    search_dirs = get_pref("RECIPE_SEARCH_DIRS") or list(DEFAULT_SEARCH_DIRS)
```
Comment (homebysix) [inline Code/autopkglib/__init__.py]: [claude-opus-4-7] Lines 745-747 duplicate `get_search_dirs()` (line 482, added in this same PR). The override branch at line 763 already calls `get_override_dirs()`. Make both sides consistent:

```python
for search_dir in get_search_dirs() + extra_search_dirs:
```

**Comment by gregneagle** [comment]
A perhaps dumb question: if we back-port the recipe map functionality from AutoPkg 3.0 to 2.x, what would then be the purpose of a 3.0? I get that 3.0 has been in development a very long time with no clear idea when it might be released; I'd think back-porting the main new feature to 2.x would make 3.0 even less likely to ever ship...

**Comment by homebysix** [comment]
@gregneagle I think we'd still bump the version to 3.0. The main blocker was that the branch used for 3.x had diverged so much from 2.x that it was not possible to unify. This PR would solve that.

**Comment by smithjw** [comment]
> Given the speed between the suggestion and PR submission I have to assume AI assistance was used. Can you share some details about what your planning approach was, what prompts you used, what AI model and effort, and any adversarial reviews or testing already performed?

@homebysix Yeah definitely. I utilised Claude Opus 4.7 within an `opencode` session to get this working. They way I structured this locally (if it's of interest), is to have the main/master branch checked out under a top-level project folder, then have each of the branches I was working with in work trees alongside it to ensure quick A/B tests across the changes and retain all the context:

```shell
~/
├─ Developer/
│  ├─ autopkg/
│  │  ├─ master/
│  │  ├─ dev-3.x/
│  │  ├─ dev-2.x/
│  │  ├─ feature-recipe-map/
```

The initial prompt was essentially to try to extract all of the recipe map commits from the `dev-3.x` branch and get them in the same working state over against `dev-2.x` but largely due to drift between the branches of multiple refactors, it would have introduced a lot of noise and conflicts while trying to merge, hence changes being recreated in the feature branch from `dev-2.x`.

After each commit/iteration of work I have the agent run all existing tests to ensure they each pass & write any new tests for new functionality. Then once green I have several pre-configured subagents that will review the work before moving onto next commits/functionality

Subagents:
- `staff-engineer` - architecture, correctness, operational safety, systemic risk, labelled blockers, & design gaps
- `junior-engineer` - clarity, unclear docs, missing context, & how easy to implement
- `security-reviewer` - vulnerabilities, threat modelling, untrusted input, auth, secrets, severity-ranked findings with remediation

After the branch was in the initially pushed status, I ran manual comparisons against the currently shipping version, dev-3.x branch, and the feature branch, prompted to go through any open issues and PRs around the recipe map functionality, and also address those items.

**Comment by nmcspadden** [comment]
Let me just say: this is great work. Thanks for picking this back up, I've had no time for years now to dedicate to this.

But I agree 100% with the others: this doesn't make sense to backport. We should instead create a new 3.0 branch forward-porting the changes and archive the previous 3.0 recipe map element. 

The changes between 2.x and the old 3.0 recipe map code are so great now that bringing 3.0 to rebase will be an invasion of monsters into a world of nightmares on the tenth circle of hell, and I wouldn't wish anyone to try.

**Comment by smithjw** [comment]
@nmcspadden are you thinking essentially, cut a brand new 3.0 branch from dev-2.x, pulling these recipe map changes there (from a commit clean up point of view), then incorporate any new 3.0 features that have been in the works and just need a polish?

Or something different?

**Comment by homebysix** [comment]
@nmcspadden I was thinking we'd essentially drop the 3.x dev branch, merge this into a new unified dev branch based on 2.x, and THEN bump the version to 3.0 so there's no more split. I'm open to whatever branching scheme you suggest, but I don't see a reason to keep the old 2.x dev branch around if it's not going to be used for releases any longer.

**Comment by eithor** [comment]
A comment from my initial testing. I also sent this to the macadmins Slack: I did a quick test and except a corner case it looks good so far. Speed is actually faster than the 3.x beta. Corner-case: if the override identifier inside a file that the recipe-map already know about gets changed (e.g. edit the file), then it fails to run the recipe because parts of the recipe uses the old identifier and corresponding paths like RECIPE_CACHE_DIR and part uses the new. The reason why it might happen in our setup is that we often copy the override to get multi-arch setup - and then change the identifier of both overrides.

**Comment by homebysix** [comment]
@eithor I've added [070eff6](https://github.com/autopkg/autopkg/pull/1027/commits/070eff6181bd8156d0302f77eac7a25740cb327b), which should address that issue. Thanks for testing and reporting!

**Review by homebysix — APPROVED** [review]
